File: app/Sidecar/SidecarStatus.php
<?php

namespace App\Sidecar;

use Carbon\CarbonImmutable;

final readonly class SidecarStatus
{
    public function __construct(
        public string $state,
        public ?string $mtgoVersion,
        public ?string $sdkVersion,
        public ?string $sidecarVersion,
        public ?string $currentFile,
        public ?CarbonImmutable $heartbeat,
        public ?int $lastEventSeq,
        public ?string $lastError,
    ) {}

    /** @param array<string, mixed> $json */
    public static function fromArray(array $json): self
    {
        return new self(
            state: (string) ($json['state'] ?? 'unknown'),
            mtgoVersion: $json['mtgo_version'] ?? null,
            sdkVersion: $json['sdk_version'] ?? null,
            sidecarVersion: $json['sidecar_version'] ?? null,
            currentFile: $json['current_file'] ?? null,
            heartbeat: isset($json['heartbeat']) ? CarbonImmutable::parse($json['heartbeat'])->utc() : null,
            lastEventSeq: isset($json['last_event_seq']) ? (int) $json['last_event_seq'] : null,
            lastError: $json['last_error'] ?? null,
        );
    }

    public function isStale(int $seconds = 10): bool
    {
        return $this->heartbeat === null || $this->heartbeat->diffInSeconds(now(), absolute: true) > $seconds;
    }

    /** @return array<string, mixed> */
    public function toArray(): array
    {
        return [
            'state' => $this->state,
            'mtgo_version' => $this->mtgoVersion,
            'sdk_version' => $this->sdkVersion,
            'sidecar_version' => $this->sidecarVersion,
            'current_file' => $this->currentFile,
            'heartbeat' => $this->heartbeat?->toIso8601ZuluString(),
            'last_event_seq' => $this->lastEventSeq,
            'last_error' => $this->lastError,
            'stale' => $this->isStale(),
        ];
    }
}

File: sidecar/MyMtgo.Sidecar.Core/Status/SidecarStatus.cs
using System.Text;
using System.Text.Json;
using MyMtgo.Sidecar.Core.Envelope;

namespace MyMtgo.Sidecar.Core.Status;

public sealed record SidecarStatus(
    int Pid,
    SidecarState State,
    string? MtgoVersion,
    string SdkVersion,
    string SidecarVersion,
    DateTimeOffset? AttachedAt,
    long LastEventSeq,
    string? CurrentFile,
    DateTimeOffset Heartbeat,
    string? LastError)
{
    public static SidecarStatus Initial(int pid, string sdkVersion, string sidecarVersion, DateTimeOffset now) =>
        new(pid, SidecarState.Waiting, null, sdkVersion, sidecarVersion, null, 0, null, now, null);

    public string ToJson()
    {
        using var stream = new MemoryStream();
        using (var w = new Utf8JsonWriter(stream))
        {
            w.WriteStartObject();
            w.WriteNumber("pid", Pid);
            w.WriteString("state", SidecarStateNames.ToWire(State));
            if (MtgoVersion is null) { w.WriteNull("mtgo_version"); } else { w.WriteString("mtgo_version", MtgoVersion); }
            w.WriteString("sdk_version", SdkVersion);
            w.WriteString("sidecar_version", SidecarVersion);
            if (AttachedAt is null) { w.WriteNull("attached_at"); } else { w.WriteString("attached_at", Timestamps.Format(AttachedAt.Value)); }
            w.WriteNumber("last_event_seq", LastEventSeq);
            if (CurrentFile is null) { w.WriteNull("current_file"); } else { w.WriteString("current_file", CurrentFile); }
            w.WriteString("heartbeat", Timestamps.Format(Heartbeat));
            if (LastError is null) { w.WriteNull("last_error"); } else { w.WriteString("last_error", LastError); }
            w.WriteEndObject();
        }

        return Encoding.UTF8.GetString(stream.ToArray());
    }
}
